lib/galaxy/objectstore/cloud.py

```diff
@@ -0,0 +1,611 @@
+                    rel_path = os.path.join(extra_dir, rel_path)
+                else:
+                    rel_path = os.path.join(rel_path, extra_dir)
+
+            # Create given directory in cache
+            cache_dir = os.path.join(self.staging_path, rel_path)
+            if not os.path.exists(cache_dir):
+                os.makedirs(cache_dir)
+
+            # Although not really necessary to create S3 folders (because S3 has
+            # flat namespace), do so for consistency with the regular file system
+            # S3 folders are marked by having trailing '/' so add it now
+            # s3_dir = '%s/' % rel_path
+            # self._push_to_os(s3_dir, from_string='')
+            # If instructed, create the dataset in cache & in S3
+            if not dir_only:
+                rel_path = os.path.join(rel_path, alt_name if alt_name else "dataset_%s.dat" % obj.id)
+                open(os.path.join(self.staging_path, rel_path), 'w').close()
+                self._push_to_os(rel_path, from_string='')
+
+    def empty(self, obj, **kwargs):
+        if self.exists(obj, **kwargs):
+            return bool(self.size(obj, **kwargs) > 0)
+        else:
+            raise ObjectNotFound( 'objectstore.empty, object does not exist: %s, kwargs: %s'
+                                 % ( str( obj ), str( kwargs ) ) )
+
+    def size(self, obj, **kwargs):
+        rel_path = self._construct_path(obj, **kwargs)
+        if self._in_cache(rel_path):
+            try:
+                return os.path.getsize(self._get_cache_path(rel_path))
+            except OSError as ex:
+                log.info("Could not get size of file '%s' in local cache, will try cloud. Error: %s", rel_path, ex)
+        elif self.exists(obj, **kwargs):
+            return self._get_size_in_cloud(rel_path)
+        log.warning("Did not find dataset '%s', returning 0 for size", rel_path)
+        return 0
+
+    def delete(self, obj, entire_dir=False, **kwargs):
+        rel_path = self._construct_path(obj, **kwargs)
+        extra_dir = kwargs.get('extra_dir', None)
+        base_dir = kwargs.get('base_dir', None)
+        dir_only = kwargs.get('dir_only', False)
+        obj_dir = kwargs.get('obj_dir', False)
+        try:
+            # Remove temparory data in JOB_WORK directory
+            if base_dir and dir_only and obj_dir:
+                shutil.rmtree(os.path.abspath(rel_path))
+                return True
+
+            # For the case of extra_files, because we don't have a reference to
+            # individual files/keys we need to remove the entire directory structure
+            # with all the files in it. This is easy for the local file system,
+            # but requires iterating through each individual key in S3 and deleing it.
+            if entire_dir and extra_dir:
+                shutil.rmtree(self._get_cache_path(rel_path))
+                results = self.bucket.list(prefix=rel_path)
+                for key in results:
```

Comment:
May be better to replace references like this that use AWS terminology to something cloud agnostic? e.g., `object`. This applies throughout the file (e.g., `_key_exists`).